### PR TITLE
Add instantiate-to-bound example to "Fixing common type problems"

### LINKED ISSUE #914: Add instantiate-to-bound example to "Fixing common type problems"
For details, see https://github.com/dart-lang/sdk/issues/33334#issuecomment-396197943.

### PR DESCRIPTION
**Pull request by chalin**
Fixes #914 

Staged at https://dartlang-org-staging-0.firebaseapp.com/guides/language/sound-problems#example-2

### CODE REVIEW

File: src/_guides/language/sound-problems.md [hunk 1]
```diff
@@ -122,6 +122,58 @@ Otherwise, use `dynamic` in situations where you cannot use a single type:
...
+While the [List][] type has an `add()` method, [Iterable][] does not.
+
+#### Fix
+
+In Dart 1.x, when a generic class is instantiated without explicit type
+arguments, `dynamic` is assumed. That is why, in the code excerpt above, `c` is
+of type `dynamic` and no error is reported for `c.add()`.
+
+In Dart 2, when a generic class is instantiated without explicit type arguments,
+each type parameter defaults to its type bound (`Interface` in this example) if
```
Comment (kwalrath) [inline src/_guides/language/sound-problems.md]: Interface -> Iterable

File: src/_guides/language/sound-problems.md [hunk 2]
```diff
@@ -66,7 +66,7 @@ error • The <member> '...' isn't defined for the class '...' • undefined_<me
 These errors usually appear in code where a variable is statically known
 to be some supertype but the code assumes a subtype.
```
Comment (kwalrath) [inline src/_guides/language/sound-problems.md]: We should alter this to talk about the case of the second example. The words we use here should be echoed in the names of the Example sections.

File: src/_guides/language/sound-problems.md [hunk 3]
```diff
@@ -122,6 +122,58 @@ Otherwise, use `dynamic` in situations where you cannot use a single type:
 var width = canvasOrImg.width;
 {% endprettify %}
 
+#### Example 2
```
Comment (kwalrath) [inline src/_guides/language/sound-problems.md]: With the second example, we need to clarify the scope of the first and second examples and their fixes. Maybe:

```
#### Example 1: <description>
##### Fix: <existing description>
#### Example 2: <description>
##### Fix: <description>
```

File: src/_guides/language/sound-problems.md (hunk 1 above)
Comment (chalin) [reply src/_guides/language/sound-problems.md]: Oops! Fixed.

**Comment by chalin** [comment]
I think that I've addressed your concerns. PTAL

File: src/_guides/language/sound-problems.md [hunk 4]
```diff
@@ -63,10 +63,7 @@ see [Runtime errors](#common-errors-and-warnings).
 error • The <member> '...' isn't defined for the class '...' • undefined_<member>
 ```
 
-These errors usually appear in code where a variable is statically known
```
Comment (kwalrath) [inline src/_guides/language/sound-problems.md]: I think we need to add back this text. Maybe make it a bulleted list.

These errors usually appear under one of the following conditions:

* A variable is statically known...
* A generic class has a bounded type parameter, and an instance of the class omits the type argument.

File: src/_guides/language/sound-problems.md [hunk 5]
```diff
@@ -153,27 +150,40 @@ error • The method 'add' isn't defined for the class 'Iterable' at lib/bounded
...
+each type parameter defaults to its type bound (`Iterable` in this example) if
 one is explicitly given, or `dynamic` otherwise.
 
 You'll need to approach fixing such errors on a case-by-case basis. It helps to
-have a good understanding of the original design intent. In this case, you might
-realize that there is a missing call to `toList()`:
+have a good understanding of the original design intent.
+
+Explicitly passing type arguments is one way to help identify type errors. For
+example, by specifying `List` as a type argument, the analyzer let's you know
```
Comment (kwalrath) [inline src/_guides/language/sound-problems.md]: There are grammar and punctuation issues in this line (grammatically, the analyzer specifies List, and there's a punctuation issue with "let's"). How about this instead:

example, if you change the code to specify `List` as a type argument, the analyzer can detect

File: src/_guides/language/sound-problems.md [hunk 6]
```diff
@@ -122,6 +119,71 @@ Otherwise, use `dynamic` in situations where you cannot use a single type:
...
+
+In Dart 1.x, when a generic class is instantiated without explicit type
+arguments, `dynamic` is assumed. That is why, in the code excerpt above, `c` is
+of type `dynamic` and no error is reported for `c.add()`.
+
+In Dart 2, when a generic class is instantiated without explicit type arguments,
+each type parameter defaults to its type bound (`Iterable` in this example) if
+one is explicitly given, or `dynamic` otherwise.
+
+You'll need to approach fixing such errors on a case-by-case basis. It helps to
```
Comment (kwalrath) [inline src/_guides/language/sound-problems.md]: You'll -> You

File: src/_guides/language/sound-problems.md [hunk 7]
```diff
@@ -122,6 +119,71 @@ Otherwise, use `dynamic` in situations where you cannot use a single type:
...
+that the constructor argument has the wrong type and needs to be changed:
+
+{:.passes-sa}
+<?code-excerpt "strong/test/strong_test.dart (add-type-arg)" replace="/.List.|\[\]/[!$&!]/g"?>
+{% prettify dart %}
+var c = new C[!<List>!]([![]!]).collection;
+c.add(2);
+{% endprettify %}
+
+If you actually meant `collection` to be an `Iterable`, then subsequent uses of
```
Comment (kwalrath) [inline src/_guides/language/sound-problems.md]: I'm confused about this. Could you clarify it?

File: src/_guides/language/sound-problems.md (hunk 4 above)
Comment (chalin) [reply src/_guides/language/sound-problems.md]: Excellent suggestion. Done

Comment (chalin) [reply src/_guides/language/sound-problems.md]: (Note that I tweak the first sentence. PTAL)

File: src/_guides/language/sound-problems.md (hunk 5 above)
Comment (chalin) [reply src/_guides/language/sound-problems.md]: Yes, much better. Done.

File: src/_guides/language/sound-problems.md (hunk 7 above)
Comment (chalin) [reply src/_guides/language/sound-problems.md]: Actually, I think that I'll just drop this (short) paragraph and its code example.

File: src/_guides/language/sound-problems.md [hunk 8]
```diff
@@ -63,6 +63,12 @@ see [Runtime errors](#common-errors-and-warnings).
 error • The <member> '...' isn't defined for the class '...' • undefined_<member>
 ```
 
+These errors can appear under the following conditions:
+
+- A variable is statically known to be some supertype but the code assumes a subtype.
```
Comment (kwalrath) [inline src/_guides/language/sound-problems.md]: but -> , but

(This was there before, but the punctuational correctness of the next item you added made me notice that this one was punctuated incorrectly. :)

File: src/_guides/language/sound-problems.md [hunk 9]
```diff
@@ -63,6 +63,12 @@ see [Runtime errors](#common-errors-and-warnings).
...
+- A generic class has a bounded type parameter, but an instance creation
+  expression of the class omits the type argument.
+
 #### Example 1: A variable is statically known to be some supertype but the code assumes a subtype
```
Comment (kwalrath) [inline src/_guides/language/sound-problems.md]: but -> , but

**Pull request merged**